### DISCUSSION TITLE
Error with up-autosubmit. Possible race condition?

### DISCUSSION THREAD

**Discussion by davisums** [unpoly/unpoly#569] [Q&A] [answered]
I'm not sure if this a bug or an error on my part, but I have a form with up-autosubmit, and from time to time, when updating a field I get a up.CannotMatch error. It doesn't happen on every autosubmit though. It seems to happen when I type "fast" and multiple updates are happening quickly after each other.

I think it might be a race condition.

Here's the html for the form. The server returns the full page each time and I'm only targeting the form using an id.

```
<form id="new_article_form" method="POST" up-autosubmit="" up-watch-delay="100" up-target="#new_article_form" up-source="/app/articles/68e17eef-9695-4b30-88de-b4fe85a07a8d" up-time="false" up-etag="false" class="up-can-clean">
      <input type="hidden" name="csrfmiddlewaretoken" value="I2hNMdyLqJ44wX1fVLMvrEe6yAPNg7441f1MYILN9Ll4veCNzTIh16yFS5FHAC8v" class="up-can-clean">
      <div id="label" class="pb-2 font-medium">
        Keyword
      </div>
      <input class="input input-bordered w-full field up-can-clean" type="text" name="keyword" value="bes" maxlength="150" x-model="keyword" required="" id="id_keyword">
      <div id="label" class="pb-2 font-medium mt-6">
        Title
      </div>
      <div class="relative">
        <input class="input input-bordered w-full field up-can-clean" type="text" name="title" value="Experience Winter Magic: The Best Ski Resorts!" maxlength="300" x-model="title" id="id_title">
      </div>
    </form>
```

<img width="1481" alt="image" src="https://github.com/unpoly/unpoly/assets/58007085/f2555418-5025-421d-a3d7-06fc1aa5ca0e">

Any idea what might be causing this?

**Comment by adam12** [comment]
For clarity, which version(s) of Unpoly are you seeing this with?

There were some changes made to Unpoly in recent(ish) versions to make the result of concurrent requests consistent.

**Reply by davisums** [reply]
Right, I forgot to add that. I tried this on 3.6.0 as well as 3.6.1 (which I'm on right now).

**Comment by triskweline** [comment]
It looks like the `[up-autosubmit]` is trying to submit the form after a debounce delay of 100ms, but the form has already been destroyed. This is strange because destroying form (e.g. through submission) should abort all watchers waiting for a debounce delay. This is one of the 3.x changes that @adam12 mentioned.

To investigate further we would need a minimal app that reproduces the issue, e.g. on Glitch.com or Codepen.

**Reply by davisums** [reply]
Here's a [small flask app](https://upautocomplete.applikuapp.com/) where I've reproduced the problem.

I can also share the repository but I'm not sure how I can turn it into a codepen since it's a get and post request both handled on the same route.

Just in case, here's all the code:

Html page:
```
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <!--Mobile friendly-->
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <script src="https://cdn.jsdelivr.net/npm/unpoly@3.6.1/unpoly.min.js"></script>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/unpoly@3.6.1/unpoly.min.css">
  </head>
<body hx-headers='{"X-CSRFToken": "zD6NQ3sX1G19VHAEkq3bu2NXj3SD8Y22HJ5DBs4Kj9ks7qCBLVeHhQWwnTPFB6Xz"}' up-main>


<div class="max-w-5xl mx-auto px-4 lg:px-8">

  <div class="container-box p-4 md:p-8" x-data="{keyword:null, title:null}">
    <form id="new_article_form" method="POST" up-target="#new_article_form" up-autosubmit up-watch-delay="100">


      <input type="hidden" name="csrfmiddlewaretoken" value="zD6NQ3sX1G19VHAEkq3bu2NXj3SD8Y22HJ5DBs4Kj9ks7qCBLVeHhQWwnTPFB6Xz">
      <div id="label" class="pb-2 font-medium">
        Keyword
      </div>
      <input class="input input-bordered w-full field field" type="text" name="keyword" maxlength="150" class="field" id="id_keyword" value="{{keyword}}">
      <div id="label" class="pb-2 font-medium mt-6">
        Title
      </div>
      <div class="relative">
        <input class="input input-bordered w-full field field" type="text" name="title" maxlength="300" class="field" id="id_title" value="{{title}}">
      </div>


    </form>
  </div>

</div>

</body>
</html>

```

Flask server:
```
from flask import Flask, render_template, request

app = Flask(__name__)


@app.route('/', methods=['GET', 'POST'])
def hello():
    if request.method == 'POST':
        keyword = request.form['keyword']
        title = request.form['title']

        return render_template('index.html', keyword=keyword, title=title)

    return render_template('index.html')

if __name__ == '__main__':
    app.run(host='0.0.0.0')

```

**Reply by adam12** [reply]
Thank you for sharing. I can confirm that I can reproduce the error by just typing fast inside one of the inputs.

Chrome
Version 120.0.6099.109 (Official Build) (arm64)

**Reply by triskweline** [reply, marked as answer]
This error will be gone in the next version (8ba0422a0c897a474b5aa4c42ad4004b6d25450d).

However, that error is just a symptom of a fundamental problem with the posted example. That form is targeting itself. Hence any user input made while the auto-submission is loading will be lost.

You should look at persisting the inputs with `[up-keep]`, or at disabling the form while loading to prevent concurrent user input (using `[up-watch-disable]`).

**Reply by davisums** [reply]
I wasn't sure what the behaviour would be (and thought the overwrite issue might be related to the error that was thrown) but that's a good point. I fixed it by adding `target=":none"`, and that also gets rid of the error.

Thank you!
